Add test nodes to CMake target "tests"

matthew-reynolds opened the pull request:
This resolves the `colcon` test failure brought up by @ipa-mdl here: https://github.com/ros-controls/ros_control/issues/410#issuecomment-574397248

See https://github.com/ros-controls/ros_control/pull/404#discussion_r366604167 for full discussion on this change. In short, targets have either got to be part of the `all` or `tests` targets, and #404 removed some targets from `all` without adding them to `tests`.

matthew-reynolds committed:
Add test nodes to CMake target "tests"

mathias-luedtke commented on combined_robot_hw_tests/CMakeLists.txt:
```diff
@@ -36,6 +36,7 @@ if(CATKIN_ENABLE_TESTING)
 
   add_executable(combined_robot_hw_dummy_app EXCLUDE_FROM_ALL src/dummy_app.cpp)
   target_link_libraries(combined_robot_hw_dummy_app ${PROJECT_NAME} ${catkin_LIBRARIES})
+  add_dependencies(tests combined_robot_hw_dummy_app)
```
I guess this should better be wrapped with `if(TARGET tests)` as in `ros_comm`.

bmagyar replied:
I'd rather move it into the `if(CATKIN_ENABLE_TESTING)` section, I strongly dislike generating more cmake blurb

mathias-luedtke replied:
> I'd rather move it into the if(CATKIN_ENABLE_TESTING)

It already is.

> I strongly dislike generating more cmake blurb

I don't understand why we moved to `ros_comm` style in the first place.
A simpler alternative is to just remove `EXCLUDE_FROM_ALL`.
The packages in question are test packages anyway.

matthew-reynolds replied:
>A simpler alternative is to just remove EXCLUDE_FROM_ALL.
The packages in question are test packages anyway.

True, but I think the extra safeguards against installing test targets is worthwhile.

So, the conclusion here is to add the `if(TARGET tests)`?

matthew-reynolds replied:
A final note: `CATKIN_ENABLE_TESTING` defaults to `ON`, so the `EXCLUDE_FROM_ALL` is quite useful for speeding up regular builds and only building tests when appropriate.

---

Back to the `if(TARGET tests)`:

> So, the conclusion here is to add the if(TARGET tests)?

I dug into this more and changed my mind, I don't think we should add this.

I don't know why `ros_comm` uses it around their `add_dependencies(tests ...)`. The Catkin CMake library will _always_ create the `tests` target ([tests.cmake L81](https://github.com/ros/catkin/blob/b6410a5257bb991481b28570899830704d768979/cmake/test/tests.cmake#L81)) unless `CATKIN_ENABLE_TESTING` is set to `OFF`. That same file also implies that it is sufficient to use only the `CATKIN_ENABLE_TESTING` check for test stuff ([tests.cmake L34](https://github.com/ros/catkin/blob/b6410a5257bb991481b28570899830704d768979/cmake/test/tests.cmake#L34)).

Additionally, the `catkin_add_gtest()` and `_gmock()` functions don't do this check before they `add_dependencies(tests ...)`; They only check for `CATKIN_ENABLE_TESTING`, and issue a warning if they are called while `CATKIN_ENABLE_TESTING` is `OFF`.

Since we are already within `if(CATKIN_ENABLE_TESTING)`, I think we are safe without `if(TARGET tests)`. If we want more clarity on why `ros_comm` uses it, we could just ask the `ros_comm` maintainers why they use the check and make our decision accordingly.

matthew-reynolds replied:
In fact, it looks like `ros_comm` even uses `if(TARGET ...)` for targets they just created the line before... See [test_rosbag L37](https://github.com/ros/ros_comm/blob/902fb00fc7b2e881575a270a99c3a077ba5cdbba/test/test_rosbag/CMakeLists.txt#L37). So their usage isn't specific to the `tests` target.

mathias-luedtke replied:
> True, but I think the extra safeguards against installing test targets is worthwhile.

We are talking about `*_test`  packages, their only purpose is to provide everything for the tests.

> Additionally, the catkin_add_gtest() and _gmock() functions don't do this check before they add_dependencies(tests ...)

[They do it for executables](https://github.com/ros/catkin/blob/dfc93c72d9799f57442d08c2f6e9fd494bba95c1/cmake/test/gtest.cmake#L82)

mathias-luedtke replied:
> In fact, it looks like ros_comm even uses if(TARGET ...) for targets they just created the line before

Yes, I guess it is best practice to check if the *external* target exists before using it.

matthew-reynolds replied:
> We are talking about *_test packages, their only purpose is to provide everything for the tests.

Indeed, but all the better if they can do so while installing as few artifacts as possible (ie. only plugins, which are required to be installed by pluginlib).

Also, while we may disagree on this point, I think the build speedup and more expected behaviour (only building these targets in test builds) still makes this worthwhile.

> [They do it for executables](https://github.com/ros/catkin/blob/dfc93c72d9799f57442d08c2f6e9fd494bba95c1/cmake/test/gtest.cmake#L82)

This line is checking that the newly created target exists (Similar to [test_rosbag L37](https://github.com/ros/ros_comm/blob/902fb00fc7b2e881575a270a99c3a077ba5cdbba/test/test_rosbag/CMakeLists.txt#L37)), not checking that the `tests` target exists, as we're debating.

mathias-luedtke reviewed (changes requested):
Thanks!

matthew-reynolds committed:
Remove dependencies on test targets

All "tests" targets will be built before any tests are run. These explicit
dependencies are neither neccessary nor standard/conventional.

mathias-luedtke reviewed (changes requested):
Please remove fc28b088a1e1ee107a1608eb3c01d7de2fd2bb21

The dependency is needed to determine the correct order, e.g. for parallel execution.
In addition the `tests`target does not get built automatically, if I am not mistaken.

matthew-reynolds commented:
>Please remove fc28b08

These targets are all run-time dependencies, not build-time, and therefore shouldn't be specified as such. From the commit msg:
```
All "tests" targets will be built before any tests are run. These explicit
dependencies are neither necessary nor standard/conventional.
```
Additionally, those build-time dependencies were only introduced in the #404 refactor - none of `ros_comm`, `ros_controllers`, or `ros_control` before #404 have it.

>The dependency is needed to determine the correct order, e.g. for parallel execution.

Every test depends on the entire `tests` target ([tests.cmake L173](https://github.com/ros/catkin/blob/b6410a5257bb991481b28570899830704d768979/cmake/test/tests.cmake#L173)) effectively forcing all tests to build before running any of them. So I don't think order and parallelization are concerns here.

> In addition the `tests` target does not get built automatically, if I am not mistaken.

The `tests` target is a dependency of the `run_tests` target, so every `run_tests` build will build these nodes. They won't be built during a normal build though, which IMO is the correct behaviour.

matthew-reynolds commented:
@ipa-mdl @bmagyar gentle bump. I'd like to fix up `colcon` builds and tie up the loose ends from #404.

matthew-reynolds commented:
Another ping @bmagyar @ipa-mdl, let's get this finished up, so I can make the mirror whatever we decide on over to `ros_controllers`.
